### PR TITLE
Add server commands

### PR DESCRIPTION
**Pull request by Jarlakxen**
Implemente https://github.com/laughedelic/atom-ide-scala/issues/27. I putted some of the code in a companion object just to keep the `ScalaLanguageClient` class small.

### CODE REVIEW

**Comment by laughedelic** [comment]
Hi @Jarlakxen. Is this work in progress?

**Comment by Jarlakxen** [comment]
Hi @laughedelic! The work is done for `clearIndexCache` and `resetPresentationCompiler`. I noticed a `scalafixUnusedImports` which is not mentioned in #27. Would you like to open a new issue? or I can try to do it all together.

**Comment by laughedelic** [comment]
Yes, it's better to add all available commands here. There's actually one more command: [`sbtConnect`](https://github.com/scalameta/metals/blob/master/vscode-extension/src/extension.ts#L108-L115).

File: .gitignore [hunk 1]
```diff
@@ -93,3 +93,4 @@ package-lock.json
 
 CHANGELOG.md
 coursier
+/bin/
```
Comment (laughedelic) [inline .gitignore]: I'm just curious, what do you have there, why do you need to ignore it in every project?

File: src/main/scala/ScalaLanguageClient.scala [hunk 2]
```diff
@@ -96,4 +96,49 @@ class ScalaLanguageClient extends AutoLanguageClient { client =>
       matching.foreach { _.dismiss() }
     }
   }
+  
+  override def postInitialization(server: ActiveServer): Unit = {
+    logger.debug(s"Server Capabilities:\n${js.JSON.stringify(server.capabilities, space = 2)}")
```
Comment (laughedelic) [inline src/main/scala/ScalaLanguageClient.scala]: Remove debug logging before merging, please. Btw, you can see it in the Atom logs anyway.

File: src/main/scala/ScalaLanguageClient.scala [hunk 3]
```diff
@@ -96,4 +96,49 @@ class ScalaLanguageClient extends AutoLanguageClient { client =>
...
+      ScalaLanguageClient.addResetPresentationCompiler(server)
+    }
+
+    if (supportedCommands.contains(ScalaLanguageClient.resetPresentationCompilerCommand)) {
+      ScalaLanguageClient.addResetPresentationCompiler(server)
+    }
+
+    if (supportedCommands.contains(ScalaLanguageClient.scalafixUnusedImportsCommand)) {
+      // TODO
+    }
```
Comment (laughedelic) [inline src/main/scala/ScalaLanguageClient.scala]: I think that all this can be drastically simplified:

**Comment by laughedelic** [comment]
I think that all this can be drastically simplified: once you got `supportedCommands`, you can just iterate over them and add each of them.

There's a bit tricky part about command names. See Atom docs:

> Command names must follow the `namespace:action` pattern, where namespace will typically be the name of your package, and action describes the behavior of your command. If either part consists of multiple words, these must be separated by hyphens. E.g. `awesome-package:turn-it-up-to-eleven`. All words should be lowercased.

* We should use either package name (`ide-scala`) or server name (`metals`) as the namespace prefix here (by the way, all this is specific to metals at the moment)
* Command name has to be transformed from camelCase to snake-case (Atom will take care of formatting it nice for the command palette)

Here's a draft:

```scala
supportedCommands.foreach { cmd =>
  val commandName = "[A-Z]".r.replaceAllIn(cmd, { "-" + _.group(0).toLowerCase() })
  Atom.commands.add("atom-text-editor", s"metals:${commandName}", { _ => 
    server.connection.executeCommand(
      new ExecuteCommandParams(command = clearIndexCacheCommand)
    )
  })
}
```

@Jarlakxen what do you think about this approach?

**Comment by Jarlakxen** [comment]
From what I understand that could work for `clearIndexCache`, `resetPresentationCompiler` and `sbtConnect` but not for `scalafixUnusedImports`. If that is correct, a solution could be define a whitelist of commands without special treatment like `clearIndexCache`, `resetPresentationCompiler` and `sbtConnect` and implement `scalafixUnusedImports` in other place.

**Comment by laughedelic** [comment]
I'm not sure I get it, why does `scalafixUnusedImports` need a special treatment?

**Comment by Jarlakxen** [comment]
I saw this part of the code https://github.com/scalameta/metals/blob/master/metaserver/src/main/scala/scala/meta/languageserver/ScalametaServices.scala#L367 and `OrganizeImports.removeUnused(params.arguments, symbolIndex)` use `params: ExecuteCommandParams`.  That `arguments` contains a `TextDocumentIdentifier`.
Maybe I didn't understand the functionality but from what I understand I need to pass a reference to the file from which I want to remove the unused imports.

**Comment by laughedelic** [comment]
I see. I think this is also solvable, but let's keep things simple for now and just skip this command: 
* the VS Code extension for metals [doesn't have](https://github.com/scalameta/metals/blob/master/vscode-extension/src/extension.ts#L91-L121) it either
* you would normally use it through the code action, so it's won't be missing

**Comment by laughedelic** [comment]
@Jarlakxen I just tried it out with the latest published server and it seems to work fine. At least I see that it receives the command request. Let's merge it!

**Pull request merged**